cleverhans/experimental/certification/nn.py

```diff
@@ -130,47 +126,36 @@ def forward_pass(self, vector, layer_index, is_transpose=False, is_abs=False):
-      elif(layer_type in {'conv', 'conv_relu'}):
-        return_vector = tf.nn.conv2d_transpose(vector, self.weights[layer_index],
+    else:
+      vector = tf.reshape(vector, self.input_shapes[layer_index])
+
+    if layer_type in {'ff', 'ff_relu'}:
+      if is_transpose:
+        weight = tf.transpose(weight)
+      return_vector = tf.matmul(weight, vector)
+    elif layer_type in {'conv', 'conv_relu'}:
+      if is_transpose:
+        return_vector = tf.nn.conv2d_transpose(vector, weight,
                                                output_shape=self.input_shapes[layer_index],
-                                               strides=[1, STRIDE, STRIDE, 1],
-                                               padding=PADDING)
+                                               strides=[1, self.stride, self.stride, 1],
+                                               padding=self.padding)
       else:
-        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
-      return tf.reshape(return_vector, (self.sizes[layer_index], 1))
-
-    elif is_abs:
-      vector = tf.reshape(vector, self.input_shapes[layer_index])
-      if(layer_type in {'ff', 'ff_relu'}):
-        return_vector = tf.matmul(tf.abs(self.weights[layer_index]), vector)
-      elif(layer_type in {'conv', 'conv_relu'}):
         return_vector = tf.nn.conv2d(vector,
-                                     tf.abs(self.weights[layer_index]),
-                                     strides=[1, STRIDE, STRIDE, 1],
-                                     padding=PADDING)
-      else:
-        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
-      return tf.reshape(return_vector, (self.sizes[layer_index + 1], 1))
-
-    # Simple forward pass
+                                     weight,
+                                     strides=[1, self.stride, self.stride, 1],
+                                     padding=self.padding)
     else:
-      vector = tf.reshape(vector, self.input_shapes[layer_index])
-      if(layer_type in {'ff', 'ff_relu'}):
-        return_vector = tf.matmul(self.weights[layer_index], vector)
-      elif (layer_type in {'conv', 'conv_relu'}):
-        return_vector = tf.nn.conv2d(vector, self.weights[layer_index],
-                                     strides=[1, STRIDE, STRIDE, 1],
-                                     padding=PADDING)
-      else:
-        raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
-      return tf.reshape(return_vector, (self.sizes[layer_index + 1], 1))
+      raise NotImplementedError('Unsupported layer type: {0}'.format(layer_type))
+    if is_transpose:
+      return tf.reshape(return_vector, (self.sizes[layer_index], 1))
+    return tf.reshape(return_vector, (self.sizes[layer_index + 1], 1))
 
-def load_network_from_checkpoint(checkpoint, model_json, input_shape=None):
+def load_network_from_checkpoint(checkpoint, model_json, input_shape=None, cnn_params=None):
```

Comment:
It would be more flexible to read cnn parameters (stride and padding) from JSON file and also make them defined individually for each layer.

Since we define format of the JSON file we can just add necessary fields (stride and padding) into conv layers.